Double click block to exit zoom out mode

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Adds ability to exit Zoom Out mode by double clicking on the block.

Note: the accessibility implications of click handlers are discussed in `Testing Instructions for Keyboard` below.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

It's important experiment with intuitive ways of exiting Zoom Out mode so that users can get back to edit mode.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Adds a double click handler to the block props which will exit zoom out mode if a block is clicked and we are in zoom out mode.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

- New Page
- Open Patterns inserter and select category to trigger Zoom Out mode
- Add some section patterns to the canvas
- Double click on any of the patterns and see that you leave Zoom Out mode.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

This action cannot be mirrored by keyboards. Therefore we have introduced an alternative means of exiting zoom out mode in https://github.com/WordPress/gutenberg/pull/64571.



## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/227347ae-44b1-4b36-8db5-7a70fee9f328

getdave committed:
Add action for exiting Zoom Out mode

getdave committed:
Add new hook for handling existing zoom out mode

MaggieCabrera commented:
I like this, and it makes sense to me. I don't see this happening by mistake easily. @richtabor I'd like your thoughts on this one.

jeryj commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -382,3 +382,14 @@ export const modifyContentLockBlock =
+ *
+ * @param {boolean} isZoomOutMode Enable/Disable Zoom Out mode.
+ */
+export const setZoomOutMode =
+	( isZoomOutMode = true ) =>
+	( { dispatch } ) => {
+		dispatch.__unstableSetEditorMode( isZoomOutMode ? 'zoom-out' : 'edit' );
+	};
```
I don't think we should add this. Exiting zoom out should return you to your previous mode, which may not be `edit`. `navigation` is a valid editor mode as well.

getdave replied:
Perhaps we can add the action but somehow get the "last" editor mode?

Noting that without a unified action creator, we're just going to end up with the same pattern littered around the code base but with different options for the "default".

Also note that the similar action we have for `setNavigationMode` uses the same pattern.

How could we improve this?

jeryj replied:
We have the `[useZoomOut()](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/hooks/use-zoom-out.js#L17)` hook that does this, but I'm unsure if it'll be reliable enough long-term. For this PR, I think it's OK to force `edit` mode as they are clicking an edit button. But I would do it via `__unstableSetEditorMode( 'zoom-out' )` directly in the [double click event handler](https://github.com/WordPress/gutenberg/pull/64573/files#diff-874ba9e21e44120b8f22a9218daaa73812555c24b91330f5122d05d8257a2641R37).

jeryj commented on packages/block-editor/src/components/block-list/use-block-props/use-zoom-out-mode-exit.js:
```diff
@@ -0,0 +1,52 @@
+					// Prevent focus from moving to the block.
+					event.preventDefault();
+
+					// When clicking on a selected block, exit navigation mode.
```
```suggestion
					// When double clicking on a selected block, exit zoom out mode.
```

jeryj commented on packages/block-editor/src/components/block-list/use-block-props/use-zoom-out-mode-exit.js:
```diff
@@ -0,0 +1,52 @@
+					event.preventDefault();
+
+					// When clicking on a selected block, exit navigation mode.
+					if ( isBlockSelected( clientId ) ) {
+						setZoomOutMode( false );
+					} else {
+						selectBlock( clientId );
+					}
```
If I double click on a non-selected block in zoom out mode, it still exits zoom out mode for me. It seems like it's due to the block being selected on the first click by another event listener.

getdave replied:
I feel like it should behave like this. However the logic I have here may need adjusting.

I want to be able to move my cursor anywhere within the canvas and double click to enter edit mode. I don't want to have to select and then double click.

How can we improve this?

jeryj replied:
I think it's already being selected on the first click by another event, so I think the `selectBlock( clientId );` else can be removed.

jeryj replied:
Yeah, this focus handler selects the block before this zoom out hook runs, so `selectBlock` in the `else` statement will never run: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/use-block-props/use-focus-handler.js#L57

getdave replied:
Updated to just go to edit directly.

jeryj reviewed (commented):
A double click to exit will be really useful!

getdave committed:
Use correct words

Co-authored-by: Jerry Jones <jones.jeremydavid@gmail.com>

getdave committed:
Remove selectBlock conditional

getdave committed:
Set edit mode directly

getdave committed:
Remove new action

jeryj commented on packages/block-editor/src/components/block-list/use-block-props/use-zoom-out-mode-exit.js:
```diff
@@ -0,0 +1,48 @@
+					event.preventDefault();
+
+					__unstableSetEditorMode( 'edit' );
+					selectBlock( clientId );
```
This should already be selected due to [the focus handler in block props selecting the block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/use-block-props/use-focus-handler.js#L57). Could we remove the `selectBlock` and leave a comment about where the block selection happens?

jeryj commented on packages/block-editor/src/components/block-list/use-block-props/use-zoom-out-mode-exit.js:
```diff
@@ -0,0 +1,48 @@
+	const { __unstableSetEditorMode, selectBlock } = unlock(
+		useDispatch( blockEditorStore )
+	);
+
```
Zoom out mode is fairly rare right now. Rather than adding a double click to all blocks, could we early return here if we're not in zoom out? That way, we only add these listeners when it's useful?

getdave commented:
@jeryj If you have time / inclination to make the changes you suggest please go ahead. Otherwise I'll pick it up during this week 🤞

jeryj committed:
Remove unused isBlockSelected prop from dependency array

jeryj committed:
Add editorMode to blockProps context to reduce useSelect and event listeners

Reducing the number of useSelect calls and event listeners is important for performance. Rather than have a useSelect to check zoom out mode and add event listeners to every block, let's do the useSelect once in the provider and pass it in to the hook, and only add the event listener if we're in zoom out mode.

jeryj committed:
Remove block selection from zoom out hook exit

The block is already being selected via the useFocusHandler hook, so no need to repeat it here.

jeryj committed:
Remove unused comment

jeryj commented:
Made some changes to reduce `useSelect` and event listeners by adding the `editorMode` to the `PrivateBlockContext`.

getdave commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -728,6 +730,7 @@ function BlockListBlockProvider( props ) {
 		themeSupportsLayout,
 		isTemporarilyEditingAsBlocks,
 		blockEditingMode,
+		editorMode,
```
As we're touching the deps, we should monitor the editor performance after merging this PR.

jeryj replied:
I was remembering https://github.com/WordPress/gutenberg/pull/56994 about combining block store subscriptions, and assumed it would be better to add it here rather than check on each block. Maybe the performance potential isn't worth it though. I do like the implementation though, as it makes the final result quite clean.

getdave replied:
That's a good point about combining store subscriptions. Let's try it "as is" and watch the performance charts post-merge.

jeryj replied:
I did some smoke test with running performance in the inspector and entering into/out of zoom out mode and this method of using the combined subscription and not adding listeners was always faster in both total scripting and rendering, so I say we go with it as well.

getdave replied:
Great. I say we merge it and then watch the performance dashboard in case of any changes post-merge.

getdave replied:
@jeryj Notice any problems in the Dashboard? I'm not seeing any.

getdave commented on packages/block-editor/src/components/block-list/use-block-props/use-zoom-out-mode-exit.js:
```diff
@@ -0,0 +1,44 @@
+ *
+ * @param {string} clientId Block client ID.
+ */
+export function useZoomOutModeExit( { editorMode } ) {
```
I'm wondering if we could just get the editor mode directly in the hook rather than pass it down as a prop? Is there a benefit to passing it down?

getdave reviewed (commented):
UX this seems to work well. 

I'm unsure if we need the changes to the block's props or whether we can just check the editor mode in the hook.

getdave commented:
@jeryj I can't approve my own PR but this feels pretty solid right now.

youknowriad commented on packages/block-editor/src/components/block-list/use-block-props/use-zoom-out-mode-exit.js:
```diff
@@ -0,0 +1,44 @@
+
+	return useRefEffect(
+		( node ) => {
+			if ( editorMode !== 'zoom-out' ) {
```
Instead of having the editorMode as a dependency of the useRefEffect and regenerating a new ref / new event listener everytime this change. We could just call getEditorMode within onDoubleClick event listener (Like we do for useNavModeExit

This would also remove the need to pass the editorMode top/down through the context...

youknowriad replied:
I'm seeing the opposite comment above. I think personally the performance impact of having the double click always there is less than re-rendering when the editorMode changes, it's also way simpler in terms of code. But I'm ok if you all disagree.

getdave replied:
I had the same thought process. I believe @jeryj did some performance profiling on this so he may be able to answer more concretely.

jeryj replied:
I had done profiling on the page and not adding the event listener used a lot less resources. The metrics for typing are going up. I'm having a hard time isolating a PR that is causing it. It's odd to me that adding the mode to the block props would cause that, as the editing mode doesn't really change often and definitely isn't changed in the middle of the performance tests. I can make another branch that reworks this to this older style and we can compare.

jeryj replied:
https://github.com/WordPress/gutenberg/pull/65326

@getdave the original way we did this also included passing `clientId` to the `refEffect` which would have made a difference in the profiling. We don't need that anymore, so hopefully the performance will be better now.

jeryj replied:
I don't see any major changes [in performance with either approach](https://github.com/WordPress/gutenberg/actions/runs/10851356209/job/30114916480?pr=65326).

Mamaduka replied:
The typing metic was probably affected by a new store subscription. See #65337.
